File: src/main/java/leetcode/editor/cn/ValidParentheses.java
package leetcode.editor.cn;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * [20] 有效的括号
 */
public class ValidParentheses {

	public static void main(String[] args) {
		Solution solution = new ValidParentheses().new Solution();
	}

	//leetcode submit region begin(Prohibit modification and deletion)
	class Solution {

		Map<Character, Character> map = new HashMap<>(6);

		{
			map.put('(', ')');
			map.put('[', ']');
			map.put('{', '}');
		}

		public boolean isValid(String s) {
			List<Character> stack = new ArrayList<>(s.length() / 2);
			for (Character c : s.toCharArray()) {
				if (isLeft(c)) {
					stack.add(c);
				} else {
					if (stack.isEmpty()) {
						return false;
					}
					Character expect = map.get(stack.remove(stack.size() - 1));
					if (expect == null || !expect.equals(c)) {
						return false;
					}
				}
			}

			return stack.isEmpty();
		}

		private boolean isLeft(Character c) {
			return map.containsKey(c);
		}
	}
//leetcode submit region end(Prohibit modification and deletion)

}

File: src/main/java/leetcode/editor/cn/ValidParentheses.go
  package leetcode.editor.cn;
  
  /**
   * [20] 有效的括号
   */
  public class ValidParentheses{
      public static void main(String[] args) {
           Solution solution = new ValidParentheses().new Solution();
      }
      //leetcode submit region begin(Prohibit modification and deletion)
func isValid(s string) bool {
  dict := make(map[byte]byte)
  dict[')'] = '('
  dict[']'] = '['
  dict['}'] = '{'
  stack := make([]byte, 0)
  for i:=0;i<len(s);i++ {
    // if current is left
    if s[i] == '(' || s[i] == '[' || s[i] == '{' {
      // push to stack
      stack = append(stack, s[i])
    } else if len(stack) == 0 {
      return false
    } else {
      // if current is right
      // check stack top if is left
      top := stack[len(stack)-1]
      if top != dict[s[i]] {
        return false
      } else {
        // pop
        stack = stack[:len(stack)-1]
      }
    }
  }
  return len(stack) == 0
}
//leetcode submit region end(Prohibit modification and deletion)

  }